crates/wasmtime/src/runtime/vm/provenance.rs

```diff
@@ -0,0 +1,215 @@
+//! list of provenances.
+//!
+//! In practice we expect that at runtime all of these provenance-related ops
+//! are noops and compile away to nothing. The only practical effect that's
+//! expected is that some optimizations may be hindered in LLVM occasionally or
+//! something like that which is by-and-large what we want to happen. Note that
+//! another practical consequence of not working with "strict provenance" means
+//! that Wasmtime is incompatible with platforms such as CHERI where exposed
+//! provenance is not available.
+
+use crate::vm::SendSyncPtr;
+use core::fmt;
+use core::ptr::NonNull;
+use core::sync::atomic::{AtomicU64, AtomicUsize};
+use wasmtime_environ::VMSharedTypeIndex;
+
+/// A pointer that is used by compiled code, or in other words is accessed
+/// outside of Rust.
+///
+/// This is intended to be the fundamental data type used to share
+/// pointers-to-things with Cranelift-compiled code for example. An example of
+/// this is that the `VMMemoryDefinition` type, which compiled code reads to
+/// learn about linear memory, uses a `VmPtr<u8>` to represent the base pointer
+/// of linear memory.
+///
+/// This type is pointer-sized and typed-like-a-pointer. This is additionally
+/// like a `NonNull<T>` in that it's never a null pointer (and
+/// `Option<VmPtr<T>>` is pointer-sized). This pointer auto-infers
+/// `Send` and `Sync` based on `T`. Note the lack of `T: ?Sized` bounds in this
+/// type additionally, meaning that it only works with sized types. That's
+/// intentional as compiled code should not be interacting with dynamically
+/// sized types in Rust.
+///
+/// This type serves two major purposes with respect to provenance and safety:
+///
+/// * Primarily this type is the only pointer type that implements `VmSafe`, the
+///   marker trait below. That forces all pointers shared with compiled code to
+///   use this type.
+///
+/// * This type represents a pointer with "exposed provenance". Once a value of
+///   this type is created the original pointer's provenance will be marked as
+///   exposed. This operation may hinder optimizations around the use of said
+///   pointer in that case.
+///
+/// This type is expected to be used not only when sending pointers to compiled
+/// code (e.g. `VMContext`) but additionally for any data at rest which shares
+/// pointers with compiled code (for example the base of linear memory or
+/// pointers stored within `VMContext` itself).
+///
+/// In general usage of this type should be minimized to only where absolutely
+/// necessary when sharing data structures with compiled code. Prefer to use
+/// `NonNull` or `SendSyncPtr` where possible.
+///
+/// > **Note**: at this time this type was just introduced to Wasmtime. This
+/// > doesn't actually do anything with provenance just yet as the original
+/// > pointer used to create a `VmPtr<T>` is preserved even at-rest. That will
+/// > change in the near future as more refactorings are completed.
```

Comment:
This comment is going to be out of date as soon as your next PR lands. Is it even worth including here? Seems like it will be easy to forget to remove down the line, might as well not even add it in the first place...